contracts/zero-ex/contracts/src/features/multiplex/BatchMultiplexFeature.sol

```diff
@@ -0,0 +1,211 @@
+        bytes[] calldata data
+    )
+        external
+        payable
+        override
+        onlyDelegateCall
+        nonReentrant(REENTRANCY_BATCH_MULTIPLEX)
+        doesNotReduceEthBalance
+        refundsAttachedEth
+        returns (bytes[] memory results)
+    {
+        results = new bytes[](data.length);
+        for (uint256 i = 0; i < data.length; i++) {
+            (bool success, bytes memory result) = address(this).delegatecall(data[i]);
+
+            if (!success) {
+                // Next 5 lines from https://ethereum.stackexchange.com/a/83577
+                if (result.length < 68) revert();
+                assembly {
+                    result := add(result, 0x04)
+                }
+                revert(abi.decode(result, (string)));
+            }
+
+            results[i] = result;
+        }
+    }
+
+    /// @inheritdoc IBatchMultiplexFeature
+    /// @notice Validator contract will assert validity of `extraData`.
+    function batchMultiplex(
+        bytes[] calldata data,
+        bytes calldata extraData,
+        address validatorAddress
+    )
+        external
+        payable
+        override
+        onlyDelegateCall
+        nonReentrant(REENTRANCY_BATCH_MULTIPLEX)
+        doesNotReduceEthBalance
+        refundsAttachedEth
+        returns (bytes[] memory results)
+    {
+        results = batchMultiplex(data, extraData, validatorAddress, ErrorHandling.REVERT);
+    }
+
+    /// @inheritdoc IBatchMultiplexFeature
+    /// @notice This method should be used to get desired behavior STOP, REVERT, CONTINUE.
+    ///   Validator contract will assert validity of `extraData`.
+    function batchMultiplex(
+        bytes[] calldata data,
+        bytes calldata extraData,
+        address validatorAddress,
+        ErrorHandling errorType
+    )
+        public
+        payable
+        override
+        onlyDelegateCall
+        nonReentrant(REENTRANCY_BATCH_MULTIPLEX)
+        doesNotReduceEthBalance
+        refundsAttachedEth
+        returns (bytes[] memory results)
+    {
+        // skip validation if validator is nil address, allows sending a batch of swaps with desired error behavior
+        //  by using nil address as validator.
+        if (validatorAddress != address(0)) {
+            _validateCalldata(data, extraData, validatorAddress);
+        }
+
+        results = new bytes[](data.length);
+        for (uint256 i = 0; i < data.length; i++) {
+            (bool success, bytes memory result) = address(this).delegatecall(data[i]);
+
+            if (!success) {
+                if (errorType == ErrorHandling.REVERT) {
+                    // Next 5 lines from https://ethereum.stackexchange.com/a/83577
+                    if (result.length < 68) revert();
+                    assembly {
+                        result := add(result, 0x04)
+                    }
+                    revert(abi.decode(result, (string)));
+                } else if (errorType == ErrorHandling.STOP) {
```

Comment:
Would it be possible that in case of error, tokens are not properly transferred back to caller? which will make it invalid to stop/continue, as it will lead to fund loss/locking

Reply:
Technically yes, for example, if a user transferred ETH and wrapped it as first call in the batch, it would be possible for the WETH in the proxy to not be returned. In this context, the client must ensure that any excess WETH is transformed back to ETH at the end of the batch, as the proxy will then refund the excess ETH. Since the method does transfer tokens to the proxy but the underlying exchange calls do that, it should not be possible for other tokens to be stuck unless that possibility was true for the underlying methods. Will try to replicate such scenario in a test.